src/compiler/checker.ts

```diff
@@ -3572,7 +3572,18 @@ module ts {
             return false;
         }
 
+        // Since removeSubtypes checks the subtype relation, and the subtype relation on a union
+        // may attempt to reduce a union, it is possible that removeSubtypes could be called
+        // recursively on the same set of types.

```

Comment:
```
// Since removeSubtypes checks the subtype relation, and the subtype relation on a union
// may attempt to reduce a union, it is possible that removeSubtypes could be called
// recursively on the same set of types. The removeSubtypesStack is used to track which
// sets of types are currently undergoing subtype reduction.
```